README.md
# Agilize

[![PyPI](https://badge.fury.io/py/Agilize.svg)](https://badge.fury.io/py/Agilize)
[![Coverage Status](https://coveralls.io/repos/github/lucasrcezimbra/agilize/badge.svg?branch=master)](https://coveralls.io/github/lucasrcezimbra/agilize?branch=master)

Unofficial client to access [Agilize](https://www.agilize.com.br/).


## Installation

```bash
pip install agilize
```


## How to Use

High-level API

```python
from agilize import Agilize, Competence


agilize = Agilize(username='11222333000160', password='p4ssw0rd')


companies = agilize.companies()

for company in companies:
    print(company)


company = companies[0]

competence = Competence(year=2022, month=5)
prolabore = company.prolabores.get(competence)
print(prolabore)

with open(f'./prolabore_{competence}.pdf', 'wb') as f:
    f.write(prolabore.download())
```

Low-level API

```python
from agilize import Client


agilize = Client(username='11222333000160', password='p4ssw0rd')

print(agilize.info)
company_id = agilize.info['party']['companies'][0]['__identity']

agilize.prolabores(company_id=company_id, year=2022)
```



## Contributing
Contributions are welcome, feel free to open an Issue or Pull Request.

Pull requests must be for the `develop` branch.

```bash
git clone https://github.com/lucasrcezimbra/agilize
cd agilize
git checkout develop
python -m venv .venv
pip install -r requirements-dev.txt
pre-commit install
pytest
```


CHANGELOG.md
# Changelog


## 0.0.10 (2022-12-14)
- Add Tax.amount and Tax.due_date


## 0.0.9 (2022-12-14)
- Add Invoice.barcode
- Update dev requirements


## 0.0.8 (2022-11-26)
- Fix upload_nfse
- Fix pre-commit
- Update dev requirements


## 0.0.7 (2022-11-08)
- Add Action to publish on PyPI [#1](https://github.com/lucasrcezimbra/agilize/issues/1)
- Fix Prolabore.from_data to cast values to Decimal
- Improve Competence adding `previous`, `next`, `last_date`, `from_date`, `first_date`
- Update dev requirements


## 0.0.6 (2022-10-29)
- Enable NFSE upload - Agilize.upload_nfse and Client.upload_nfse
- Update dev requirements


## 0.0.5 (2022-07-23)
- Download invoices NFSE
- Update dev dependencies
  * faker 13.15.0 ~> 13.15.1


## 0.0.4 (2022-07-15)
- BREAKING CHANGE: Rename Client.download_paycheck to download_prolabore
- BREAKING CHANGE: Update Client.PATH_*
- Add taxes with download
- Add Client.invoices
- Add script to help fixture generation
- Update dev dependencies
  * faker 13.14.0 ~> 13.15.0
  * pre-commit 2.19.0 ~> 2.20.0
  * pytest-mock 3.8.1 ~> 3.8.2


## 0.0.3 (2022-06-25)
- Add pro-labore download
  * add Client.download_paycheck
- Create high-level API
- Update requirements-dev


## 0.0.2 (2022-04-17)
- First version - authentication and prolabores


tests/conftest.py
import random
from uuid import uuid4

import pytest
from faker.providers.python import Provider as PythonProvider

from agilize import Client, Company


class Provider(PythonProvider):
    def money(self, negative=False):
        return self.pyfloat(
            left_digits=random.randint(3, 4),
            right_digits=2,
            positive=(not negative),
        )


@pytest.fixture(scope='session', autouse=True)
def faker_session_locale():
    return ['pt_BR']


@pytest.fixture(autouse=True)
def faker(faker):
    faker.add_provider(Provider)
    return faker


@pytest.fixture
def client_mock(mocker):
    return mocker.create_autospec(Client)


@pytest.fixture
def company(faker, client_mock):
    return Company(
        id=str(uuid4()),
        cnpj='',
        name=faker.company(),
        client=client_mock,
    )


@pytest.fixture
def info_data(faker):
    return {
        'accountIdentifier': '12345678000160',
        'creationDate': '2022-04-17T17:11:00-0300',
        'expirationDate': None,
        'party': {
            'companies': [
                {
                    'activityType': 1,
                    'city': {'code': '1234567', 'name': 'Porto Alegre'},
                    'clientSince': '2022-04-17T17:11:00-0300',
                    'cnpj': '12345678000160',
                    'email': faker.email(),
                    'firstEmail': faker.email(),
                    'foundingDate': '2022-04-17T17:13:00-0300',
                    'hasEmployees': False,
                    'hasFinanceiro': True,
                    'isAgilizePremium': False,
                    'isBlocked': False,
                    'isComercio': False,
                    'isHabilitadoEmitirNfse': False,
                    'isLucroPresumido': False,
                    'isMei': False,
                    'isOperadoPorProcuracao': False,
                    'lockedAt': None,
                    'name': faker.company(),
                    '__identity': str(uuid4()),
                }
            ],
            'email': faker.email(),
            'emailIsVerified': True,
            'name': None,
            'temporaryEmail': None,
            '__identity': str(uuid4()),
        },
        'roles': {
            'Agilize.Api:Customer': {
                'identifier': 'Agilize.Api:Customer',
                'name': 'Customer',
            }
        },
    }


@pytest.fixture
def company_data(info_data):
    return info_data['party']['companies'][0]


@pytest.fixture
def partners_data(faker):
    return [
        {
            "amountOfDependent": faker.pyint(),
            "birthDate": "1900-09-03T00:00:00-0300",
            "company": {"isAutomaticFatorR": faker.pybool()},
            "currentProlabore": faker.pyint(),
            "hasIncompleteDataForEsocial": faker.pybool(),
            "hasUnregisteredDependent": faker.pybool(),
            "isAdmin": faker.pybool(),
            "isSeniorPartner": faker.pybool(),
            "mainDocument": "01234567890",
            "name": faker.name(),
            "person": {
                "birthDate": "1900-09-03T00:00:00-0300",
                "certificatesA3": [],
                "cityOfBirth": {
                    "__isInitialized__": faker.pybool(),
                    "agilizeEnabled": faker.pybool(),
                    "code": "1234567",
                    "codigoSIAFI": "1234",
                    "dddCode": "55",
                    "name": "Porto Alegre",
                    "slugName": "porto_alegre",
                    "timezone": None,
                    "__identity": str(uuid4()),
                },
                "countryOfBirth": None,
                "countryOfNationality": None,
                "cpf": "01234567890",
                "defaultAddress": {
                    "addressType": None,
                    "area": None,
                    "city": {
                        "__isInitialized__": faker.pybool(),
                        "agilizeEnabled": faker.pybool(),
                        "code": "1234567",
                        "codigoSIAFI": "1234",
                        "dddCode": "55",
                        "name": "Porto Alegre",
                        "slugName": "porto_alegre",
                        "timezone": None,
                        "__identity": str(uuid4()),
                    },
                    "cityCode": "1234567",
                    "cityName": "Porto Alegre",
                    "complement": "house 1",
                    "formattedAddress": "Rua A, 123,Complemento: house 1",
                    "formattedForProlabore": "Rua A, 123,Complemento: house 1",
                    "invalid": faker.pybool(),
                    "invalidReason": None,
                    "iptu": None,
                    "isDefault": faker.pybool(),
                    "neighborhood": "São Paulo",
                    "number": "123",
                    "originalPlainText": None,
                    "semNumero": faker.pybool(),
                    "state": {
                        "__isInitialized__": faker.pybool(),
                        "abbreviation": "RS",
                        "code": "12",
                        "name": "Rio Grande do Sul",
                        "__identity": str(uuid4()),
                    },
                    "stateAbbreviation": "RS",
                    "street": "Rua A",
                    "zipCode": "90000000",
                    "Flow_Persistence_clone": faker.pybool(),
                    "__identity": str(uuid4()),
                },
                "dependencias": [],
                "documents": [
                    {
                        "documentNumber": "12345678910",
                        "documentType": {
                            "__isInitialized__": faker.pybool(),
                            "abbreviation": "PIS",
                            "identification": faker.pybool(),
                            "mask": None,
                            "personType": 1,
                            "slugname": "pispasep",
                            "__identity": str(uuid4()),
                        },
                        "__identity": str(uuid4()),
                    }
                ],
                "foreignAddress": None,
                "foreignRegisterExpedition": None,
                "foreignRegisterNumber": "",
                "foreignRegisterOrganizationEmmit": "",
                "gender": faker.pyint(),
                "hasChildrenWithBrazilian": None,
                "hasForeignRegister": None,
                "isVerificadoNaRfb": faker.pybool(),
                "marriedWithBrazilian": None,
                "name": faker.name(),
                "origin": 1,
                "pisPasep": "12345678910",
                "relationWithBrazil": None,
                "stateOfBirth": {
                    "__isInitialized__": faker.pybool(),
                    "abbreviation": "RS",
                    "code": "12",
                    "name": "Rio Grande do Sul",
                    "__identity": str(uuid4()),
                },
                "verificadoNaRfb": faker.pybool(),
                "__meta": {"isPartnerDataComplete": faker.pybool()},
                "__identity": str(uuid4()),
            },
            "salaryAmount": 0,
            "share": faker.pyint(),
            "updatedAt": "2022-04-17T21:56:00-0300",
            "__identity": str(uuid4()),
        }
    ]


@pytest.fixture
def prolabores_data(faker):
    return {
        "2022-02-01": [
            {
                "amountOfDependent": 0,
                "competence": "2022-02-01T00:00:00-0300",
                "contraCheque": {"__identity": str(uuid4())},
                "deducaoDependente": 0,
                "iNSS": faker.money(),
                "iNSSBrl": str(faker.money()),
                "iNSSPatronal": 0,
                "iNSSPatronalBrl": "0,00",
                "iNSSTotal": faker.money(),
                "iNSSTotalBrl": str(faker.money()),
                "iRPJFolha": 0,
                "iRPJFolhaBrl": "0,00",
                "iRRF": False,
                "iRRFAliquota": 0,
                "iRRFAliquotaAsPercent": 0,
                "iRRFBase": faker.money(),
                "isAnexoIV": False,
                "isLucroPresumido": False,
                "isSimplesNacional": True,
                "pagamentoProlabore": None,
                "partner": {"__identity": str(uuid4())},
                "partnerName": faker.name(),
                "provisaoINSSPatronal": None,
                "rubricasDescontos": [],
                "rubricasProventos": [],
                "salaryAmount": 0,
                "saldoInssAPagarAposCompensacoes": faker.money(),
                "totalCompensacoes": 0,
                "totalDeImpostos": faker.money(),
                "valor": faker.pyint(),
                "valorBrl": str(faker.money()),
                "valorLiquido": faker.money(),
                "valorLiquidoBrl": str(faker.money()),
                "valorTotalDescontos": 0,
                "valorTotalProventos": 0,
            }
        ],
        "2022-01-01": [
            {
                "amountOfDependent": 0,
                "competence": "2022-01-01T00:00:00-0300",
                "contraCheque": {"__identity": str(uuid4())},
                "deducaoDependente": 0,
                "iNSS": faker.money(),
                "iNSSBrl": str(faker.money()),
                "iNSSPatronal": 0,
                "iNSSPatronalBrl": "0,00",
                "iNSSTotal": faker.money(),
                "iNSSTotalBrl": str(faker.money()),
                "iRPJFolha": 0,
                "iRPJFolhaBrl": "0,00",
                "iRRF": False,
                "iRRFAliquota": 0,
                "iRRFAliquotaAsPercent": 0,
                "iRRFBase": faker.money(),
                "isAnexoIV": False,
                "isLucroPresumido": False,
                "isSimplesNacional": True,
                "pagamentoProlabore": None,
                "partner": {"__identity": str(uuid4())},
                "partnerName": faker.name(),
                "provisaoINSSPatronal": None,
                "rubricasDescontos": [],
                "rubricasProventos": [],
                "salaryAmount": 0,
                "saldoInssAPagarAposCompensacoes": faker.money(),
                "totalCompensacoes": 0,
                "totalDeImpostos": faker.money(),
                "valor": faker.pyint(),
                "valorBrl": str(faker.money()),
                "valorLiquido": faker.money(),
                "valorLiquidoBrl": str(faker.money()),
                "valorTotalDescontos": 0,
                "valorTotalProventos": 0,
            }
        ],
    }


@pytest.fixture
def taxes_data(faker):
    return [
        {
            'automaticAuditErrorCode': '500',
            'billet': {'__identity': str(uuid4())},
            'billetLastUpdate': '2022-07-03T12:34:56-0300',
            'closed': faker.pybool(),
            'competence': '2022-06-01T00:00:00-0300',
            'competenceIntervalDate': {
                'from': '2022-06-01T00:00:00-0300',
                'to': '2022-06-30T23:59:59-0300',
            },
            'createdBy': None,
            'deadline': '2022-07-20T23:59:59-0300',
            'dueTax': faker.money(),
            'dueTaxLessThanMinValue': faker.pybool(),
            'evidence': None,
            'evidenceLastUpdate': None,
            'finalPaid': faker.pybool(),
            'hasCertified': None,
            'hasHonorarioSucumbencia': faker.pybool(),
            'interest': None,
            'isAutomaticRecalculation': faker.pybool(),
            'lastRecalculationAt': None,
            'manuallyCreatedReason': None,
            'needRecalculation': faker.pybool(),
            'newDeadlineRequest': None,
            'officialDueValue': faker.money(),
            'officialEvidences': [],
            'paid': faker.pybool(),
            'paidAt': None,
            'penalties': None,
            'periodRevenue': faker.money(),
            'recalculation': faker.pybool(),
            'releasedToClient': faker.pybool(),
            'retention': 0,
            'showTax': faker.pybool(),
            'status': 1,
            'tax': {
                'abbreviation': 'DAS',
                'disabledAt': None,
                'friendlyName': None,
                'name': 'Documento de Arrecadação do Simples Nacional',
                'slugName': 'DAS',
                'taxType': 3,
                '__identity': str(uuid4()),
            },
            'taxAbbreviation': 'DAS',
            'taxMinValueToPay': 10,
            'taxSlugName': 'DAS',
            'taxWillAggregatePayment': None,
            'total': faker.money(),
            'underReview': faker.pybool(),
            'updatedAt': '2022-07-02T13:48:19-0300',
            'valorFaturamento': 0,
            'valorImposto': faker.money(),
            'valorImpostos': faker.money(),
            'valorRetencoes': 0,
            'wasManuallyCreated': faker.pybool(),
            '__identity': str(uuid4()),
        },
    ]


@pytest.fixture
def invoices_data(faker):
    return [
        {
            '__identity': str(uuid4()),
            'closingDate': '2022-07-10T04:00:00-0300',
            'code': f'{faker.company_id()}-082022',
            'competence': '2022-07-01T00:00:00-0300',
            'deadline': '2022-07-15T23:59:59-0300',
            'entries': [
                {
                    '__identity': str(uuid4()),
                    'amount': faker.money(),
                    'amountComDesconto': None,
                    'description': faker.sentence(),
                    'isNegative': False,
                    'keyInvoiceEntry': f'{uuid4()}_mensalidade',
                    'mensalidade': True,
                    'type': 0,
                    'weight': -1,
                },
                {
                    '__identity': str(uuid4()),
                    'amount': faker.money(),
                    'amountComDesconto': None,
                    'description': faker.sentence(),
                    'isNegative': False,
                    'keyInvoiceEntry': str(uuid4()),
                    'mensalidade': False,
                    'type': 0,
                    'weight': 2,
                },
            ],
            'nfses': [
                {
                    '__identity': str(uuid4()),
                    'aliquota': 0.05,
                    'anexoSimples': None,
                    'anexoSimplesSlugname': None,
                    'atividadeMunicipal': None,
                    'atividadeMunicipalCode': None,
                    'avaliacao': 2,
                    'baseCalculo': faker.money(),
                    'basePresuncaoIrpj': 0.12,
                    'canBeCanceled': faker.pybool(),
                    'canSendEmailToPrestador': faker.pybool(),
                    'canSendEmailToTomador': faker.pybool(),
                    'cancelDenyReason': None,
                    'canceled': faker.pybool(),
                    'cityActivity': None,
                    'cityCode': str(faker.pyint()),
                    'codigoVerificacao': faker.pystr(),
                    'contabilDataProcessedAt': None,
                    'customIr': None,
                    'descontoCondicionado': 0,
                    'descontoIncondicionado': 0,
                    'discriminacao': faker.sentence(),
                    'economicActivity': None,
                    'email': faker.email(),
                    'emailed': None,
                    'emailedToPrestador': None,
                    'emitted': faker.pybool(),
                    'encodedXml': faker.pystr(),
                    'environment': 1,
                    'fatorR': 0,
                    'fiscalDataProcessedAt': None,
                    'fonte': 1,
                    'hasBeenEmailed': faker.pybool(),
                    'hasBeenEmailedToPrestador': faker.pybool(),
                    'hasCancellation': faker.pybool(),
                    'hasEvaluation': faker.pybool(),
                    'hasEvaluations': faker.pybool(),
                    'inscricaoMunicipal': None,
                    'irpjRetido': faker.pybool(),
                    'isInssRetido': faker.pybool(),
                    'isNotaExterior': faker.pybool(),
                    'isProcessed': faker.pybool(),
                    'isSimplesNacional': faker.pybool(),
                    'isTomadorSameCity': faker.pybool(),
                    'issRetido': faker.pybool(),
                    'itemListaServico': None,
                    'motivoRejeicao': None,
                    'naturezaOperacao': 1,
                    'nfseAuthUrl': faker.url(),
                    'nfseKey': faker.pystr(),
                    'nfseUrl': (
                        'https://nfse.salvador.ba.gov.br/site/contribuinte/nota/notaprint.aspx'
                        f'?nf={faker.pyint()}'
                        f'&inscricao={faker.pyint()}'
                        f'&verificacao={faker.pystr()}'
                    ),
                    'numero': str(faker.pyint()),
                    'numeroAndSerieDescription': f'{faker.pyint()} (Série: )',
                    'numeroSerie': str(faker.pyint()),
                    'outrasRetencoes': 0,
                    'pisCofinsCsllRetido': faker.pybool(),
                    'pristineAnexoSimples': None,
                    'pristineAnexoSimplesSlugName': None,
                    'processed': None,
                    'realBaseCalculo': faker.money(),
                    'regimeEspecialTributacao': 0,
                    'rejeitada': faker.pybool(),
                    'relatedTaxesProcessedAt': None,
                    'serie': None,
                    'servicoPrestadoExterior': faker.pybool(),
                    'situacao': 1,
                    'situacaoUpdatedAt': None,
                    'tomadorName': faker.company(),
                    'totalDueTax': faker.money(),
                    'totalRecebido': faker.money(),
                    'totalRetido': 0,
                    'totalTributos': faker.money(),
                    'valorCofins': 0,
                    'valorCofinsRatio': 1.23,
                    'valorCofinsTotal': faker.money(),
                    'valorCsll': 0,
                    'valorCsllTotal': faker.money(),
                    'valorCustomIr': 0,
                    'valorDeducoes': 0,
                    'valorInss': 0,
                    'valorInssTotal': 0,
                    'valorIr': 0,
                    'valorIrTotal': faker.money(),
                    'valorIss': 0,
                    'valorIssTotal': faker.money(),
                    'valorLiquido': faker.money(),
                    'valorPis': 0,
                    'valorPisRatio': 0.12,
                    'valorPisTotal': faker.money(),
                    'valorServicos': faker.money(),
                    'valorSimplesNacional': 0,
                    'xml': faker.pystr(),
                }
            ],
            'overdue': faker.pybool(),
            'paidPaymentOrder': {
                '__identity': str(uuid4()),
                'splitItems': [],
            },
            'paymentDate': '2022-07-10T00:00:00-0300',
            'status': faker.pyint(),
            'total': faker.money(),
        },
    ]


@pytest.fixture
def invoice_payment_data(faker):
    amount = faker.money()
    barcode = faker.pystr_format("###############################################")
    return {
            "closingDate": "2022-12-10T04:00:00-0300",
            "code": "{CNPJ}-122022",
            "competence": "2022-12-01T00:00:00-0300",
            "deadline": "2022-12-15T23:59:59-0300",
            "entries": [
                {
                    "amount": amount,
                    "amountComDesconto": None,
                    "description": "Mensalidade da Agilize relativa à competência...",
                    "isNegative": faker.pybool(),
                    "keyInvoiceEntry": f"{uuid4()}_mensalidade",
                    "mensalidade": faker.pybool(),
                    "type": 0,
                    "weight": -1,
                    "__identity": str(uuid4()),
                },
            ],
            "overdue": faker.pybool(),
            "paidPaymentOrder": None,
            "paymentDate": None,
            "paymentOrders": [
                {
                    "qrCode": faker.url(),
                    "qrCodeCopyPaste": faker.pystr(),
                    "__identity": str(uuid4()),
                },
                {
                    "billet": {
                        "barcode": barcode,
                        "chargeTotal": amount,
                        "isExpired": faker.pybool(),
                        "url": f"https://faturas.iugu.com/{uuid4()}.pdf",
                        "__identity": str(uuid4()),
                    },
                    "__identity": str(uuid4()),
                },
            ],
            "status": 1,
            "total": amount,
            "validBilletPaymentOrder": {
                "billet": {
                    "barcode": barcode,
                    "chargeTotal": amount,
                    "isExpired": faker.pybool(),
                    "url": f"https://faturas.iugu.com/{uuid4()}.pdf",
                    "__identity": str(uuid4()),
                },
                "createdAt": "2022-12-10T04:21:45-0300",
                "isReadyForPayment": faker.pybool(),
                "__identity": str(uuid4()),
            },
            "__identity": str(uuid4()),
        }


tests/test_client.py
from uuid import uuid4

import pytest
import responses
from responses import matchers

from agilize import Client
from agilize.client import URL
from agilize.keycloak import Keycloak


@pytest.fixture
def client(mocker):
    return Client('john', 'p4ssw0rd', keycloak=mocker.create_autospec(Keycloak))


def test_init(mocker):
    keycloak_mock = mocker.patch('agilize.client.Keycloak', autospec=True)

    client = Client('john', 'p4ssw0rd')

    assert client.username == 'john'
    assert client.password == 'p4ssw0rd'
    keycloak_mock.assert_called_once_with(
        Client.AUTH_URL, Client.CLIENT_ID, Client.REALM_NAME
    )


def test_access_token(client, mocker):
    access_token = 'random_string'
    client.keycloak.token.return_value = {'access_token': access_token}

    assert client.access_token == access_token
    client.keycloak.token.assert_called_once_with(client.username, client.password)


def test_cache_access_token(client):
    client.access_token
    client.keycloak.token.assert_called_once()

    client.keycloak.token.reset_mock()
    client.access_token
    client.keycloak.token.assert_not_called()


@responses.activate
def test_info(client, info_data):
    responses.add(
        responses.GET,
        URL.INFO,
        json=info_data,
        match=[
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.info == info_data


@responses.activate
def test_cache_info(client, info_data):
    responses.add(responses.GET, URL.INFO, json=info_data)

    client.info
    client.info

    responses.assert_call_count(URL.INFO, 1)


@responses.activate
def test_prolabores(client, prolabores_data):
    company_id, year = uuid4(), 2022

    responses.add(
        responses.GET,
        URL.PROLABORE.format(company_id=company_id),
        json=prolabores_data,
        match=[
            matchers.query_string_matcher(f'anoReferencia={year}-01-01T00:00:00P'),
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.prolabores(company_id, year) == prolabores_data


@responses.activate
def test_download_prolabore(client, prolabores_data):
    company_id, partner_id, year, month = uuid4(), uuid4(), 2022, 3
    file = b''

    responses.add(
        responses.GET,
        URL.DOWNLOAD_PROLABORE.format(company_id=company_id),
        body=file,
        match=[
            matchers.query_string_matcher(
                f'competence={year}-{month}-01T00:00:00-0300&partner={partner_id}'
            ),
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.download_prolabore(company_id, partner_id, year, month) == file


@responses.activate
def test_partners(client, partners_data):
    company_id = uuid4()

    responses.add(
        responses.GET,
        URL.PARTNERS.format(company_id=company_id),
        json=partners_data,
        match=[
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.partners(company_id) == partners_data


@responses.activate
def test_taxes(client, taxes_data):
    company_id, year = uuid4(), 2022

    responses.add(
        responses.GET,
        URL.TAXES.format(company_id=company_id),
        json=taxes_data,
        match=[
            matchers.query_param_matcher(
                {
                    'blocking': True,
                    'closed': True,
                    'count': 3000,
                    'direction': 'desc',
                    'onlyTaxesNotProvisionedByRh': True,
                    'page': 1,
                    'sort': 'companyTax.competence',
                    'year': year,
                },
            ),
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.taxes(company_id, year) == taxes_data


@responses.activate
def test_invoices(client, invoices_data):
    company_id, year = uuid4(), 2022

    responses.add(
        responses.GET,
        URL.INVOICES.format(company_id=company_id),
        json=invoices_data,
        match=[
            matchers.query_param_matcher({'count': 3000, 'page': 1, 'year': year}),
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.invoices(company_id, year) == invoices_data


@responses.activate
def test_download_tax(client, faker, prolabores_data):
    company_id, tax_id = uuid4(), uuid4()
    file_url = faker.url()
    file = b''

    responses.add(
        responses.GET,
        URL.DOWNLOAD_TAX.format(company_id=company_id, tax_id=tax_id),
        json={'url': file_url},
        match=[
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )
    responses.add(responses.GET, file_url, body=file)

    assert client.download_tax(company_id, tax_id) == file


@responses.activate
def test_upload_invoice(client, faker):
    company_id = uuid4()
    file = b''
    file_identity = str(uuid4())
    expected_response = {
        'countNfses': 1,
        'createdAt': '2022-10-29T14:40:00-0300',
        'nfses': [],
    }

    responses.add(
        responses.POST,
        URL.UPLOAD_NFSE.format(company_id=company_id),
        json={'__identity': file_identity},
        match=[
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"}),
            matchers.multipart_matcher({'resources[0]': ('whatever.xml', file, 'text/xml')}),
        ],
    )
    responses.add(
        responses.POST,
        URL.UPLOAD_NFSE2.format(company_id=company_id),
        json=expected_response,
        match=[
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"}),
            matchers.json_params_matcher({"nfseImportResource": file_identity}),
        ],
    )

    assert client.upload_nfse(company_id, file) == expected_response


@responses.activate
def test_invoice_payment(client, invoice_payment_data):
    company_id, invoice_id = uuid4(), uuid4()

    responses.add(
        responses.GET,
        URL.INVOICE_PAYMENT.format(company_id=company_id, invoice_id=invoice_id),
        json=invoice_payment_data,
        match=[
            matchers.header_matcher({"Authorization": f"Bearer {client.access_token}"})
        ],
    )

    assert client.invoice_payment(company_id, invoice_id) == invoice_payment_data


agilize/client.py
import requests

from agilize.keycloak import Keycloak


class URL:
    BASE = 'https://app.agilize.com.br/api/v1/'

    DOWNLOAD_PROLABORE = BASE + 'companies/{company_id}/prolabore-anual/download'
    DOWNLOAD_TAX = BASE + 'companies/{company_id}/taxes/{tax_id}/billet'
    INFO = BASE + 'companies/security-user/info'
    INVOICES = BASE + 'companies/{company_id}/invoices'
    INVOICE_PAYMENT = BASE + 'companies/{company_id}/invoices/{invoice_id}'
    PARTNERS = BASE + 'companies/{company_id}/partners'
    PROLABORE = BASE + 'companies/{company_id}/prolabore-anual'
    TAXES = BASE + 'companies/{company_id}/taxes'
    UPLOAD_NFSE = BASE + 'companies/{company_id}/nfseimportresources'
    UPLOAD_NFSE2 = BASE + 'companies/{company_id}/nfses/importfromresource'


class AnonymousClient:
    @staticmethod
    def download(url):
        return requests.get(url).content


class Client(AnonymousClient):
    AUTH_URL = 'https://sso.agilize.com.br/auth/'
    CLIENT_ID = 'agilize-legacy-client'
    REALM_NAME = 'AgilizeAPPs'

    def __init__(self, username, password, keycloak=None):
        self.username = username
        self.password = password
        self.keycloak = keycloak or Keycloak(self.AUTH_URL, self.CLIENT_ID, self.REALM_NAME)
        self._access_token = None
        self._info = None

    @property
    def access_token(self):
        if not self._access_token:
            token = self.keycloak.token(self.username, self.password)
            self._access_token = token['access_token']
        return self._access_token

    @property
    def info(self):
        if not self._info:
            response = requests.get(
                url=URL.INFO,
                headers=self.headers,
            )
            self._info = response.json()
        return self._info

    @property
    def headers(self):
        return {'Authorization': f'Bearer {self.access_token}'}

    def partners(self, company_id):
        response = requests.get(
            url=URL.PARTNERS.format(company_id=company_id),
            headers=self.headers,
        )
        return response.json()

    def prolabores(self, company_id, year):
        response = requests.get(
            url=URL.PROLABORE.format(company_id=company_id),
            params={'anoReferencia': f'{year}-01-01T00:00:00P'},
            headers=self.headers,
        )
        return response.json()

    def download_prolabore(self, company_id, partner_id, year, month):
        response = requests.get(
            url=URL.DOWNLOAD_PROLABORE.format(company_id=company_id),
            params={
                'competence': f'{year}-{month}-01T00:00:00-0300',
                'partner': partner_id,
            },
            headers=self.headers,
        )
        return response.content

    def taxes(self, company_id, year):
        response = requests.get(
            url=URL.TAXES.format(company_id=company_id),
            params={
                'blocking': True,
                'closed': True,
                'count': 3000,
                'direction': 'desc',
                'onlyTaxesNotProvisionedByRh': True,
                'page': 1,
                'sort': 'companyTax.competence',
                'year': year,
            },
            headers=self.headers,
        )
        return response.json()

    def download_tax(self, company_id, tax_id):
        response = requests.get(
            url=URL.DOWNLOAD_TAX.format(company_id=company_id, tax_id=tax_id),
            headers=self.headers,
        )
        return self.download(response.json()['url'])

    def invoices(self, company_id, year):
        response = requests.get(
            url=URL.INVOICES.format(company_id=company_id),
            params={
                'count': 3000,
                'page': 1,
                'year': year,
            },
            headers=self.headers,
        )
        return response.json()

    def invoice_payment(self, company_id, invoice_id):
        response = requests.get(
            url=URL.INVOICE_PAYMENT.format(company_id=company_id, invoice_id=invoice_id),
            headers=self.headers,
        )
        return response.json()

    def upload_nfse(self, company_id, filebytes):
        response = requests.post(
            url=URL.UPLOAD_NFSE.format(company_id=company_id),
            files={'resources[0]': ('whatever.xml', filebytes, 'text/xml')},
            headers=self.headers,
        )
        response2 = requests.post(
            url=URL.UPLOAD_NFSE2.format(company_id=company_id),
            json={'nfseImportResource': response.json()['__identity']},
            headers=self.headers,
        )
        return response2.json()
